### PR TITLE
Add *WithIndex instances for StrMap

### PR DESCRIPTION
**Pull request by Rufflewind**
Add `FunctorWithIndex`, `FoldableWithIndex`, and `TraversableWithIndex` instances for `StrMap`.

### CODE REVIEW

File: src/Data/StrMap.purs [hunk 1]
```diff
@@ -112,10 +118,20 @@ instance foldableStrMap :: Foldable StrMap where
   foldr f z m = foldr f z (values m)
   foldMap f = foldMap (const f)
 
+instance foldableWithIndex :: FoldableWithIndex String StrMap where
+  foldlWithIndex f = fold (flip f)
+  foldrWithIndex f z m = foldr (uncurry f) z (toArrayWithKey Tuple m)
+  foldMapWithIndex = foldMap
```
Comment (paf31) [inline src/Data/StrMap.purs]: How does this one work?

File: src/Data/StrMap.purs [hunk 2]
```diff
@@ -112,10 +118,20 @@ instance foldableStrMap :: Foldable StrMap where
...
+
 instance traversableStrMap :: Traversable StrMap where
   traverse f ms = fold (\acc k v -> insert k <$> f v <*> acc) (pure empty) ms
   sequence = traverse id
 
+instance traversableWithIndexStrMap :: TraversableWithIndex String StrMap where
+  traverseWithIndex f ms = fold (\acc k v -> insert k <$> f k v <*> acc)
```
Comment (paf31) [inline src/Data/StrMap.purs]: Can this be defined in terms of `sequence` and `mapWithKey`? It doesn't have to be, just a thought.

File: src/Data/StrMap.purs (hunk 1 above)
Comment (Rufflewind) [inline src/Data/StrMap.purs]: - `foldlWithIndex`: nearly equivalent to `fold`.  The `flip` converts `k -> acc -> v -> acc` to `k -> v -> acc -> acc`.  (The ordering
- `foldrWithIndex`: first unpack into an array of `Tuple k v`, then use the standard `foldr` for arrays.  The `uncurry` converts `k -> v -> acc -> acc` to `Tuple k v -> acc -> acc` suitable for `foldr`.
- `foldMapWithIndex`: same as `foldMap`.

File: src/Data/StrMap.purs (hunk 2 above)
Comment (Rufflewind) [inline src/Data/StrMap.purs]: Apparently yes! :P

File: src/Data/StrMap.purs (hunk 1 above)
Comment (paf31) [inline src/Data/StrMap.purs]: I meant `foldMap` specifically. I wasn't seeing how the types lined up. Could you please add a test?

**Comment by Rufflewind** [comment]
Um, so I tried adding a test that relates `foldMapWithIndex f` to `traverseWithIndex (\k v -> tell (f k v))`, but `traverseWithIndex` kept giving me the reverse of what I expect.

I think `traverse` is broken:

```purescript
> runWriter (traverse_ tell ["1", "2", "3"])
(Tuple unit "123")
> runWriter (traverse_ tell (StrMap.fromFoldable [Tuple "a" "1", Tuple "b" "2", Tuple "c" "3"]))
(Tuple unit "321")
> fold (StrMap.fromFoldable [Tuple "a" "1", Tuple "b" "2", Tuple "c" "3"])
"123"
```

This seems to fix it:

```diff
-  traverse f ms = fold (\acc k v -> insert k <$> f v <*> acc) (pure empty) ms
+  traverse f ms = fold (\acc k v -> flip (insert k) <$> acc <*> f v) (pure empty) ms
```

I'm also a bit wary of `sequence <<< mapWithKey` as well.  It seems to work, but the object is being *recreated* and JavaScript doesn’t necessarily guarantee the object is insertion-ordered, so this could break on some browsers. It might be better to just define `traverse` in terms of `traverseWithIndex` instead.

**Comment by paf31** [comment]
> It might be better to just define traverse in terms of traverseWithIndex instead.

That seems fine to me.

Anyone else have comments here?

File: src/Data/StrMap.purs [hunk 3]
```diff
@@ -112,10 +118,19 @@ instance foldableStrMap :: Foldable StrMap where
   foldr f z m = foldr f z (values m)
   foldMap f = foldMap (const f)
 
+instance foldableWithIndex :: FoldableWithIndex String StrMap where
```
Comment (hdgarrood) [inline src/Data/StrMap.purs]: `foldableWithIndexStrMap`?

**Comment by paf31** [comment]
I don't know if it's broken, but a `Traversable` container should be ordered, so we need to pick some ordering, even if it's just alphabetical by key.

**Comment by hdgarrood** [comment]
At the moment I think it's whatever order the JS engine decides to use in a for-in loop, which is usually insertion order, not by key - but the order used in a for-in loop is not actually defined in the language spec AFAIK.

For example:
```purescript
> entries = [Tuple "a" 1, Tuple "c" 3, Tuple "b" 2]
> m = SM.fromFoldable entries
> traverse_ logShow m
1
3
2
```

**Comment by paf31** [comment]
Seems like we could fix that easily enough by iterating over the sorted keys, right?

**Comment by hdgarrood** [comment]
Perhaps, but it's potentially quite a large performance cost. I don't see that there's a problem with the current `Traversable` behaviour or the behaviour in this PR, personally.

**Comment by hdgarrood** [comment]
If you want fast lookup, insert, & delete, and a traversal with a guaranteed order, you should be using `Map String a`, I think.

**Comment by natefaubion** [comment]
Maybe we should deprecate StrMap, and move it to a JavaScript specific module, with the caveat that it matches JS iteration semantics.

**Comment by paf31** [comment]
Obviously `StrMap` is unordered. But `Foldable` says it is ordered. Does that mean we can use `fromFoldable :: _ ~> Array` to observe the insertion order? That sounds like a bug to me.

**Comment by hdgarrood** [comment]
I'm not sure I'd agree `StrMap` is unordered; in my mind it's for JS objects which are being used like maps, and in particular where every property is of a fixed type. So a key part of the API in my mind is that a `StrMap` has this runtime representation.

JS objects necessarily have an ordering, in that a for-in loop has to iterate over them in some order, even if that order is unspecified and therefore unpredictable. It's maybe not ideal, but given that this is how they work, I think it would be strange for `StrMap` to do anything other than exhibit the same behaviour.

For the same reason, I'm not sure we should go as far as deprecating `StrMap`; it can be useful for interop, in particular. (Having said that, it would make me happier to see more `Map String a` usage where it is more appropriate than `StrMap`, which imo is most of the time.) It might make sense to move `StrMap` to a separate package but I'm not sure it would be worth the breakage.

**Comment by natefaubion** [comment]
I mean deprecating it as a part of purescript-maps. It's clearly useful, but it isn't a general purpose data structure. It only exists as a hack for JS interop. if you wrote FFI bindings for a different backend, would you expect to keep all the same JS semantics?

**Comment by natefaubion** [comment]
Also, it's frustrating that the API for it and Map are not kept in sync. People use it for very different use cases.

**Comment by hdgarrood** [comment]
Yes, that's a good point. If we were going to do this we should take the opportunity to rename it too, perhaps JSMap or something.

**Comment by hdgarrood** [comment]
Anyway we should probably move this to a separate issue at this point.

**Comment by Rufflewind** [comment]
> `foldableWithIndexStrMap`?

Oops, should be fixed now.

**Comment by hdgarrood** [comment]
Great, thanks. So should we merge this, make a minor release, and open a new issue to decide what to do about `StrMap`?

**Comment by paf31** [comment]
[Here is a demonstration of the bug](http://try.purescript.org/?gist=bcd1d35c8c058b11561d791b02b21447)

**Pull request merged**